sparse/impl/KokkosSparse_CrsMatrix_traversal_impl.hpp

```diff
@@ -0,0 +1,173 @@
+  KOKKOS_INLINE_FUNCTION void operator()(const ordinal_type rowIdx) const {
+    for (size_type entryIdx = A.graph.row_map(rowIdx);
+         entryIdx < A.graph.row_map(rowIdx + 1); ++entryIdx) {
+      const ordinal_type colIdx = A.graph.entries(entryIdx);
+      const value_type value    = A.values(entryIdx);
+
+      func(rowIdx, entryIdx, colIdx, value);
+    }
+  }
+
+  // TeamPolicy overload
+  KOKKOS_INLINE_FUNCTION void operator()(const team_member_type& dev) const {
+    const ordinal_type teamWork = dev.league_rank() * rows_per_team;
+    Kokkos::parallel_for(
+        Kokkos::TeamThreadRange(dev, rows_per_team), [&](ordinal_type loop) {
+          // iRow represents a row of the matrix, so its correct type is
+          // ordinal_type.
+          const ordinal_type rowIdx = teamWork + loop;
+          if (rowIdx >= A.numRows()) {
+            return;
+          }
+
+          const ordinal_type row_length =
+              A.graph.row_map(rowIdx + 1) - A.graph.row_map(rowIdx);
+          Kokkos::parallel_for(
+              Kokkos::ThreadVectorRange(dev, row_length),
+              [&](ordinal_type rowEntryIdx) {
+                const size_type entryIdx = A.graph.row_map(rowIdx) +
+                                           static_cast<size_type>(rowEntryIdx);
+                const ordinal_type colIdx = A.graph.entries(entryIdx);
+                const value_type value    = A.values(entryIdx);
+
+                func(rowIdx, entryIdx, colIdx, value);
+              });
+        });
+  }
+};
+
+template <class execution_space>
+int64_t crsmatrix_traversal_launch_parameters(int64_t numRows, int64_t nnz,
+                                              int64_t rows_per_thread,
+                                              int& team_size,
+                                              int& vector_length) {
+  int64_t rows_per_team;
+  int64_t nnz_per_row = nnz / numRows;
+
+  if (nnz_per_row < 1) nnz_per_row = 1;
+
+  int max_vector_length = 1;
+#ifdef KOKKOS_ENABLE_CUDA
+  if (std::is_same<execution_space, Kokkos::Cuda>::value)
+    max_vector_length = 32;
+#endif
+#ifdef KOKKOS_ENABLE_HIP
+  if (std::is_same<execution_space, Kokkos::Experimental::HIP>::value)
+    max_vector_length = 64;
+#endif
+
+  if (vector_length < 1) {
+    vector_length = 1;
+    while (vector_length < max_vector_length && vector_length * 6 < nnz_per_row)
+      vector_length *= 2;
+  }
+
+  // Determine rows per thread
+  if (rows_per_thread < 1) {
+    if (KokkosKernels::Impl::kk_is_gpu_exec_space<execution_space>())
+      rows_per_thread = 1;
+    else {
+      if (nnz_per_row < 20 && nnz > 5000000) {
+        rows_per_thread = 256;
+      } else
+        rows_per_thread = 64;
+    }
+  }
+
+  if (team_size < 1) {
+    if (KokkosKernels::Impl::kk_is_gpu_exec_space<execution_space>()) {
+      team_size = 256 / vector_length;
+    } else {
+      team_size = 1;
+    }
+  }
+
+  rows_per_team = rows_per_thread * team_size;
+
+  if (rows_per_team < 0) {
```

Comment:
Can ``rows_per_team`` ever be negative here? Looks like both ``rows_per_thread`` and ``team_size`` must both be positive